utils/udp_client.go

```diff
@@ -35,41 +37,95 @@ type AgentClientUDP struct {
 	agent.Agent
 	io.Closer
 
-	connUDP       *net.UDPConn
+	connUDP       udpConn
 	client        *agent.AgentClient
 	maxPacketSize int                   // max size of datagram in bytes
 	thriftBuffer  *thrift.TMemoryBuffer // buffer used to calculate byte size of a span
 }
 
-// NewAgentClientUDP creates a client that sends spans to Jaeger Agent over UDP.
-func NewAgentClientUDP(hostPort string, maxPacketSize int) (*AgentClientUDP, error) {
-	if maxPacketSize == 0 {
-		maxPacketSize = UDPPacketMaxLength
+type udpConn interface {
+	Write([]byte) (int, error)
+	SetWriteBuffer(int) error
+	Close() error
+}
+
+// AgentClientUDPParams allows specifying options for initializing an AgentClientUDP. An instance of this struct should
+// be passed to NewAgentClientUDPWithParams.
+type AgentClientUDPParams struct {
+	HostPort      string
+	MaxPacketSize int
+	Logger        log.Logger
+	ResolveFunc   resolveFunc
+	DialFunc      dialFunc
+}
+
+// NewAgentClientUDPWithParams creates a client that sends spans to Jaeger Agent over UDP.
+func NewAgentClientUDPWithParams(params AgentClientUDPParams) (*AgentClientUDP, error) {
+	if params.MaxPacketSize == 0 {
+		params.MaxPacketSize = UDPPacketMaxLength
 	}
 
-	thriftBuffer := thrift.NewTMemoryBufferLen(maxPacketSize)
+	if params.Logger == nil {
+		params.Logger = log.StdLogger
+	}
+
+	if params.ResolveFunc == nil {
+		params.ResolveFunc = net.ResolveUDPAddr
+	}
+
+	if params.DialFunc == nil {
+		params.DialFunc = net.DialUDP
+	}
+
+	thriftBuffer := thrift.NewTMemoryBufferLen(params.MaxPacketSize)
 	protocolFactory := thrift.NewTCompactProtocolFactory()
 	client := agent.NewAgentClientFactory(thriftBuffer, protocolFactory)
 
-	destAddr, err := net.ResolveUDPAddr("udp", hostPort)
+	var connUDP udpConn
+	var err error
+
+	host, _, err := net.SplitHostPort(params.HostPort)
 	if err != nil {
 		return nil, err
 	}
 
-	connUDP, err := net.DialUDP(destAddr.Network(), nil, destAddr)
-	if err != nil {
-		return nil, err
+	if ip := net.ParseIP(host); ip == nil {
```

Comment:
Should there be some fallback logic for `localhost`? Otherwise it will always go with reconnecting version.

An even bigger question - do we really need this auto-detection capability? It's always a slippery slope. I think I would rather have a config parameter that lets user opt-in for the reconnecting version. This would make `utils/udp_client_test.go` much simpler, we'd only need to test which connection the client selects, and the connection tests are already in `utils/resolved_udp_conn_test.go`.

Reply:
Yep true localhost should definitely not be handled by the re-connector. In my opinion the auto-detection capability seems reasonable since as a user of this library i did not expect the issues resolved by this pr. Maybe it could be an opt-out option?

Reply:
My concern with auto-detection is that (a) it changes the current behavior, and (b) it's hard to anticipate how it will behave in all scenarios.

I agree that people probably use this lib more often in environments where agent is referenced by a host name than where it is `localhost` or fixed IP.

What if, instead of auto-detect, it would default to reconnecting version, and have an opt-out mode for fixed connection? When fixed IP is given, the reconnecting version is mostly only different in using an RWLock for each Write (fortunately, R-lock is a spin lock, and the sender is running in a single go-routine, so it should be very cheap).

BTW, please expose reconnect interval as a parameter.

Reply:
If we go with opt-in and the user provides an ip for host should it still choose the auto resolved type?